File: me/austin/queer/module/hacks/Hack.java
package me.austin.queer.module.hacks;

import me.austin.queer.module.Module;
import me.austin.queer.module.setting.Setting;
import me.austin.queer.module.setting.Settings;
import me.austin.queer.module.setting.settings.KeyBindSetting;
import me.austin.queer.util.Util;
import net.minecraft.network.Packet;

public abstract class Hack extends Module implements Util {
	public static Hack INSTANCE;
	private boolean enabled = false;
	private final KeyBindSetting bind;
	private final Category category;

	public Hack(String name, String description, int bind, Category category) {
		super(name, description);
		this.bind = new KeyBindSetting(bind, this);
		this.category = category;
	}
	
	public void onEnable() {
	}
	
	public void onDisable() {
	}

	public void onUpdate() {
	}

	public void onPacketRecieve(Packet<?> packet) {
	}
	
	public void enable() {
		if (!this.enabled) {
			this.enabled = true;
			this.onEnable();
		}
	}
	
	public void disable() {
		if (this.enabled) {
			this.enabled = false;
			this.onDisable();
		}
	}
	
	public void toggle() {
		if (this.enabled) {
			this.disable();
		} else {
			this.onEnable();
		}
		this.enabled = !this.enabled;
	}
	
	public KeyBindSetting getBind() {
		return this.bind;
	}
	
	public void setBind(int bind) {
		this.bind.set(bind);
	}
	
	public boolean isEnabled() {
		return this.enabled;
	}
	
	public Category getCategory() {
		return this.category;
	}

	public static Setting<?> register(Setting<?> setting) {
		Settings.add(setting);
		return setting;
	}
}

File: src/main/kotlin/trans/rights/client/modules/hack/Hack.kt
package trans.rights.client.modules.hack

import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import trans.rights.client.TransRights.Companion.LOGGER
import trans.rights.client.manager.impl.HackManager
import trans.rights.client.modules.Module
import trans.rights.client.modules.setting.Settings
import trans.rights.client.modules.setting.impl.BooleanSetting
import trans.rights.client.modules.setting.impl.EnumSetting
import trans.rights.client.modules.setting.impl.NumberSetting
import trans.rights.client.util.file.FileHelper
import trans.rights.event.bus.impl.BasicEventManager
import java.io.File

abstract class Hack(
    name: String,
    description: String,
    val settings: Settings = Settings(),
    val file: File = File(HackManager.directory.absolutePath + "$name.json"),
    private var enabled: Boolean = false
) : Module(name, description) {

    fun enable() {
        if (this.enabled) return

        BasicEventManager.register(this)

        this.onEnable()
        this.enabled = true
    }

    protected fun disable() {
        if (!this.enabled) return

        BasicEventManager.unregister(this)

        this.onDisable()
        this.enabled = false
    }

    fun toggle() {
        if (this.enabled) disable() else enable()
    }

    open fun onEnable() {}

    open fun onDisable() {}

    // TODO: Make this better / more streamlined
    fun load(file: File) {
        try {
            if (!file.exists()) file.createNewFile()

            if (FileHelper.read(this.file.toPath()) == "") {
                this.save(file)
                return
            }

            val json = FileHelper.fromJson(file.toPath())

            this.enabled = json.get("enabled").asBoolean

            for (setting in settings.values) {
                when (setting) {
                    is BooleanSetting -> setting.value = json.get(setting.name).asBoolean
                    is NumberSetting -> setting.value = json.get(setting.name).asDouble
                }
            }
        } catch (e: Exception) {
            FileHelper.clearJson(file.toPath())

            LOGGER.error("$name failed to load")

            e.printStackTrace()
        }
    }

    fun save(file: File) {
        try {
            val json = JsonObject()

            json.add("enabled", JsonPrimitive(enabled))

            for (setting in settings.values) {
                when (setting) {
                    is BooleanSetting -> json.add(setting.name, JsonPrimitive(setting.value))
                    is NumberSetting -> json.add(setting.name, JsonPrimitive(setting.value))
                    is EnumSetting -> json.add(setting.name, JsonPrimitive(setting.value.toString()))
                }
            }

            FileHelper.writeToJson(json, file.toPath())
        } catch (e: Exception) {
            LOGGER.error("$name failed to save")

            e.printStackTrace()
        }
    }
}
